src/khepri.erl

```diff
@@ -412,1297 +480,1623 @@ get_store_ids() ->
-%% will point to an empty map, indicating there was no existing node (i.e.
-%% there was no properties or payload to return).
-%%
-%% The payload must be one of the following form:
-%% <ul>
-%% <li>An explicit absence of payload ({@link khepri_payload:no_payload()}),
-%% using the marker returned by {@link khepri_payload:none/0}, meaning there
-%% will be no payload attached to the node and the existing payload will be
-%% discarded if any</li>
-%% <li>An anonymous function; it will be considered a stored procedure and will
-%% be wrapped in a {@link khepri_payload:sproc()} record</li>
-%% <li>Any other term; it will be wrapped in a {@link khepri_payload:data()}
-%% record</li>
-%% </ul>
+%% before the tree node is looked up: so if a condition in the path could
+%% potentially match several nodes, an exception is raised, even though only
+%% one tree node would match at the time.
 %%
-%% It is possible to wrap the payload in its internal structure explicitly
-%% using the {@link khepri_payload} module directly.
+%% The returned `{ok, Payload}' tuple contains the payload of the targetted
+%% tree node, or `{ok, undefined}' if the tree node had no payload.
 %%
-%% The `Extra' map may specify put-specific options:
-%% <ul>
-%% <li>`keep_while': `keep_while' conditions to tie the life of the inserted
-%% node to conditions on other nodes; see {@link
-%% khepri_condition:keep_while()}.</li>
-%% </ul>
+%% Example: query a tree node which hold the atom `value'
+%% ```
+%% %% Query the tree node at `/:foo/:bar'.
+%% {ok, value} = khepri:get(StoreId, [foo, bar]).
+%% '''
 %%
-%% The `Options' map may specify command-level options; see {@link
-%% command_options()}.
+%% Example: query an existing tree node with no payload
+%% ```
+%% %% Query the tree node at `/:no_payload'.
+%% {ok, undefined} = khepri:get(StoreId, [no_payload]).
+%% '''
 %%
-%% Example:
+%% Example: query a non-existent tree node
 %% ```
-%% %% Insert a node at `/:foo/:bar', overwriting the previous value.
-%% Result = khepri:put(StoreId, [foo, bar], new_value),
-%%
-%% %% Here is the content of `Result'.
-%% {ok, #{[foo, bar] => #{data => old_value,
-%%                        payload_version => 1,
-%%                        child_list_version => 1,
-%%                        child_list_length => 0}}} = Result.
+%% %% Query the tree node at `/:non_existent'.
+%% {error, ?khepri_error(node_not_found, _)} = khepri:get(
+%%                                               StoreId, [non_existent]).
 %% '''
 %%
 %% @param StoreId the name of the Khepri store.
-%% @param PathPattern the path (or path pattern) to the node to create or
-%%        modify.
-%% @param Data the Erlang term or function to store, or a {@link
-%%        khepri_payload:payload()} structure.
-%% @param Extra extra options such as `keep_while' conditions.
-%% @param Options command options such as the command type.
+%% @param PathPattern the path (or path pattern) to the tree node to get.
```

Comment:
No because patterns can be used to match a specific tree node. For instance, you can include conditions on the payload, its version, the mere existence of the tree node and so on.

What is forbidden is the use of conditions that could match many tree nodes (like a regex on the tree node's name) or more generally that prevent the code to determine the exact node of the tree node.

Reply:
Ah good point, thanks 👍